docs/design/seccomp.md

```diff
@@ -0,0 +1,380 @@
+
+  // The default action that should be taken if there are no rules defined for a system call.
+	DefaultAction SeccompAction `json:"defaultAction"`
+  // The rules for this profile.
+  Rules []SeccompRule `json:"rules,omitempty"`
+}
+
+// A seccomp syscall filter rule.
+type SeccompRule struct {
+	// The name of the system call to use
+	Name string `json:"name"`
+	// The action to take
+	Action Action `json:"action"`
+	// The arguments to the filter
+	Args []*Arg `json:"args"`
+}
+
+// An action associated with a seccomp rule.
+type SeccompAction string
+
+const (
+  // The Kill action results in the triggering task being killed if the system call is made.
+	Kill SeccompAction = "Kill"
+  // The Trap action results in the kernel sending a SIGSYS signal to the triggering task
+  // without executing the system call.
+  Trap  SeccompAction = "Trap"
+  // The Errno action results in the lower 16 bits of the return value being passed to
+  // userland as the errno without executing the system call.
+  Errno SeccompAction = "Errno"
+  // The Trap action results in kernel attempting to notify a ptrace()-based tracer prior
+  // to executing the system call.  If there is no tracer present, ENOSYS is returned to
+  // userland and the system call is not executed.
+	Trace SeccompAction = "Trace"
+  // The Allow action results in the system call being executed.
+  Allow SeccompAction = "Allow"
+)
+
+// A comparison operator to be used when matching syscall arguments in Seccomp
+type SeccompOperator string
+
+const (
+	EqualTo              SeccompOperator = "EqualTo"
+	NotEqualTo           SeccompOperator = "NotEqualTo"
+	GreaterThan          SeccompOperator = "GreaterThan"
+	GreaterThanOrEqualTo SeccompOperator = "GreaterThanOrEqualTo"
+	LessThan             SeccompOperator = "LessThan"
+	LessThanOrEqualTo    SeccompOperator = "LessThanOrEqualTo"
+	MaskEqualTo          SeccompOperator = "MaskEqualTo"
+)
+
+// Represents a specific syscall argument in a seccomp rule.
+type SeccompArg struct {
+	Index    int             `json:"index"`
+	Value    int64           `json:"value"`
+	ValueTwo int64           `json:"valueTwo"`
+	Op       SeccompOperator `json:"op"`
+}
+```
+
+Along with this API change, there will be supporting code changes:
+
+1.  New REST API implementation
+2.  Client code
+2.  Validations/conversions
+3.  `kubectl` support (including describer)
+
+### Pod API changes
+
+The container `securityContext` must be changed to allow specifying a seccomp profile:
+
+```go
+// SecurityContext holds security configuration that will be applied to a container.
+// Some fields are present in both SecurityContext and PodSecurityContext.  When both
+// are set, the values in SecurityContext take precedence.
+type SecurityContext struct {
+  // other fields omitted
+  // SeccompProfileName is the name of the seccomp profile this pod should run under.
+  SeccompProfileName string `json:"seccompProfileName,omitempty"`
```

Comment:
So in docker we were going to implement the server api like this, but the thing is if you are running the client on 1 machine and the server on another, then you need to make sure the profile actually exists on the path on the server machine, unsure how this applies to kubernetes wrt client vs server but depending on which server node you hit it kinda assumes it needs to be there. Which in a massive cluster might be hard.
We ended up reading the profile through the client and passing the struct through the api